src/cascadia/TerminalControl/ControlCore.cpp

```diff
@@ -1753,6 +1753,39 @@ namespace winrt::Microsoft::Terminal::Control::implementation
         return hstring{ str };
     }
 
+    // Get all of our recent commands. This will only really work if the user has enabled shell integration.
+    Control::CommandHistoryContext ControlCore::CommandHistory() const
+    {
+        auto terminalLock = _terminal->LockForWriting();
+        auto context = winrt::make_self<CommandHistoryContext>();
+        const auto& textBuffer = _terminal->GetTextBuffer();
+
+        for (const auto& mark : _terminal->GetScrollMarks())
+        {
+            // The command text is between the `end` (which denotes the end of
+            // the prompt) and the `commandEnd`.
+            bool markHasCommand = mark.commandEnd.has_value() &&
+                                  mark.commandEnd != mark.end;
+            if (!markHasCommand)
+                continue;
+
+            // Get the text of the command
+            const auto line = mark.end.y;
+            const auto& row = textBuffer.GetRowByOffset(line);
+            const auto rowText = row.GetText();
+            const auto commandText = rowText.substr(mark.end.x, mark.commandEnd->x - mark.end.x);

```

Comment:
@lhecker This is the original place where I should be doing something better. Something kinda like what you're doing in `ROW::CopyRangeFrom`? Like

```c++
auto ROW::GetText(til::CoordType columnBegin, til::CoordType columnLimit)
{
    const auto otherColBeg = other._clampedColumnInclusive(otherBegin);
    const auto otherColLimit = other._clampedColumnInclusive(otherLimit);
    std::span<uint16_t> charOffsets;
    std::wstring_view chars;

    if (otherColBeg < otherColLimit)
    {
        charOffsets = other._charOffsets.subspan(otherColBeg, static_cast<size_t>(otherColLimit) - otherColBeg + 1);
        const auto charsOffset = charOffsets.front() & CharOffsetsMask;
        // We _are_ using span. But C++ decided that string_view and span aren't convertible.
        // _chars is a std::span for performance and because it refers to raw, shared memory.
#pragma warning(suppress : 26481) // Don't use pointer arithmetic. Use span instead (bounds.1).
        chars = { other._chars.data() + charsOffset, other._chars.size() - charsOffset };
    }
    
    return chars;
}

```

?

Reply:
That's fairly close actually, but you can simplify it by using the `_uncheckedCharOffset` helpers which handle the `CharOffsetsMask` for you. I pushed 2 commits into this PR which contain such a `GetText` function: https://github.com/microsoft/terminal/pull/15027